src/demikernel/dbuf.rs

```diff
@@ -0,0 +1,189 @@
+            return Ok(data_ptr);
+        }
+
+        Err(Fail::new(libc::EINVAL, "zero-length buffer"))
+    }
+}
+
+//==============================================================================
+// Standard-Library Trait Implementations
+//==============================================================================
+
+/// Partial Equality Trait Implementation for Data Buffers
+impl PartialEq for DataBuffer {
+    fn eq(&self, rhs: &Self) -> bool {
+        self[..] == rhs[..]
+    }
+}
+
+/// Equality Trait Implementation for Data Buffers
+impl Eq for DataBuffer {}
+
+/// Conversion Trait Implementation for Data Buffers
+impl From<&[u8]> for DataBuffer {
+    fn from(src: &[u8]) -> Self {
+        let buf: Arc<[u8]> = src.into();
+        Self {
+            data: Some(buf),
+            offset: 0,
+            len: src.len(),
+        }
+    }
+}
+
+/// Debug Trait Implementation for Data Buffers
+impl Debug for DataBuffer {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "DataBuffer({:?})", &self[..])
+    }
+}
+
+/// De-Reference Trait Implementation for Data Buffers
+impl Deref for DataBuffer {
+    type Target = [u8];
+
+    fn deref(&self) -> &[u8] {
+        match self.data {
+            None => &[],
+            Some(ref buf) => &buf[self.offset..(self.offset + self.len)],
+        }
+    }
+}
+
+/// Mutable De-Reference Trait Implementation for Data Buffers
+impl DerefMut for DataBuffer {
+    fn deref_mut(&mut self) -> &mut [u8] {
+        match self.data {
+            None => &mut [],
+            Some(ref mut data) => {
+                let slice: &mut [u8] = Arc::get_mut(data).expect("cannot write to a shared buffer");
+                &mut slice[self.offset..(self.offset + self.len)]
+            },
+        }
+    }
+}
+
+//==============================================================================
+// Buffer Trait Implementation
+//==============================================================================
+
+/// Buffer Trait Implementation for Data Buffers
+impl Buffer for DataBuffer {
+    /// Creates an empty buffer.
+    fn empty() -> Self {
+        Self {
+            data: None,
+            offset: 0,
+            len: 0,
+        }
+    }
+
+    /// Creates a data buffer from a slice.
+    fn from_slice(src: &[u8]) -> Self {
+        src.into()
+    }
+
+    /// Removes bytes from the front of the target data buffer.
+    fn adjust(&mut self, nbytes: usize) {
+        if nbytes > self.len {
+            panic!("adjusting past end of buffer: {} vs {}", nbytes, self.len);
+        }
+        self.offset += nbytes;
+        self.len -= nbytes;
+    }
+
+    /// Removes bytes from the end of the target data buffer.
+    fn trim(&mut self, nbytes: usize) {
+        if nbytes > self.len {
+            panic!(
+                "trimming past beginning of buffer: {} vs {}",
+                nbytes, self.len
+            );
```

Comment:
This comment is more applicable to your formatting PR.

Ideally, this (lines 182 to 185) would all be on one line, like for line 173.  But it looks like this is formatting for 80-character lines.  I haven't used a punch-card machine in over 40 years, or even a VT-100 or VT-220 in over 30-something years.  I think we can reasonably go to 120-character lines now.  But if this was still too long for one line, this formatting is still wrong, as line 184 should be two lines, one each for "nbytes," and "self.len".

Reply:
This is not about an 80-character line configuration, and neither of your suggestions work.

It is imposed by the rust formatting style applied to the whole source tree. It imposes a 100-character maximum width: 

https://github.com/demikernel/demikernel/blob/master/rustfmt.toml#L1

Note that changing this parameter will affect the entire code base.

Reply:
Ah, that's the file we need to fix!  But that isn't really relevant to this PR.

Reply:
Let's discuss this formatting issue on our weekly meeting.